### ISSUE TITLE
nObject class and functions for it.

### ISSUE THREAD

**Issue by Smilebags** [Smilebags/p5.dimensions.js#22]
This is a place to track proposals and progress on the nObject class and related functions. I'm working on an initial constructor function to lay out the basics of how it will all work.

I'm thinking of moving to typescript for this to allow for easier development.

**Comment by Smilebags** [comment]
@max0410 are you familiar with TypeScript and able to work with it? If so I might work on re-organizing the code as TS files then bring them together compile-time.

**Comment by max0410** [comment]
Yes, I am familiar with typescript. I might be a bit rusty though. But, isn't just javascript with classes and some other fancy functions? I wouldn't mind if you did it in typescript.

**Comment by max0410** [comment]
I added nShift #23

**Comment by Smilebags** [comment]
Thanks for the PR.

Typescript (as you probably know) allows strong typing. It means we can define data structures for things like the nObject, which can then be error-checked and viewed anywhere in the code. It helps things to be more readable and scalable, and means we can easily split the library into more manageable chunks. If you have a suitable environment for working in TS, I'll convert the project.

**Comment by max0410** [comment]
Yes, i have an environment.

I recently got @shiffman 's attention on a live stream about p5js dimensions, and he said he would gladly do it if I either tweet or sent a github issue on his channel repo with more info. I know you'll probably say yes, but I just want to get your approval.

**Comment by Smilebags** [comment]
Great.

For him to convert it to TS? Or to publicise it? Thanks for asking but you don't need my permission to do stuff like that! Go ahead! This is our repo, you've done half the work, so feel free to do anything you like.

**Comment by Smilebags** [comment]
![image](https://cloud.githubusercontent.com/assets/24507835/23088422/5ed44308-f5d0-11e6-946b-ba71639ceb43.png)
^^

**Comment by max0410** [comment]
No, no, no! He would just make a video about it! He wouldn't be making a cool project out of it,  not contributing.

**Comment by max0410** [comment]
I guess I just though you'd like to know.

**Comment by Smilebags** [comment]
Yeah, that sounds good! Thanks for letting me know. Maybe I'll create a new branch for TS conversion (and we will work on the nObject in there?) so that the library isn't constantly changing, potentially with bugs to anyone trying to use it.

**Comment by max0410** [comment]
Ok!

**Comment by Smilebags** [comment]
I have had a look at converting to TS and it is more messy than I thought. I mean all you _have_ to do is rename the file, but to get it to work nicely and not show lots of errors is a bit harder. Are you familiar with how to say that the project is dependent on p5 in TS?

**Comment by Smilebags** [comment]
I have a feeling some sort of import should work but I just don't know how to get it to work.

**Comment by max0410** [comment]
To tell the truth I don't know much about TS. I'll research it.

**Comment by Smilebags** [comment]
Me too! *high-5*   I know it is worth using for larger projects but I've only really dabbled in the basics of it.

**Comment by max0410** [comment]
Do you really get any errors just by renaming the file? Probably some warnings, but I don''t see why there would be errors.

**Comment by max0410** [comment]
Also I sent him a tweet: 
@shiffman Here is something for a coding train video! Try out p5 dimensions! http://github.com/Smilebags/p5.dimensions.js

Hopefully he responds soon!

**Comment by Smilebags** [comment]
TypeScript Errors are more picky than JavaScript ones. For example every p5.prototype.something line gives an error saying "Cannot find name 'p5'."

**Comment by Smilebags** [comment]
It still compiles, if you allow it to, but it isn't clean TypeScript.

**Comment by Smilebags** [comment]
Thanks! I look forward to his reply.

**Comment by max0410** [comment]
Ah

**Comment by max0410** [comment]
About, the nD viewer idea. Here is a site about 4d visualizations: http://www.math.union.edu/~dpvc/math/4d/models/welcome.html

**Comment by Smilebags** [comment]
This is the same idea I had, pretty much. But I want to make it easy to interact with. It would be perfect if rotating the object in nD and moving/controlling the camera in nD was simple. Also the 'slice' view they show, I've been trying to work out how to do it. Doesn't seem easy. Cool link, thanks!

**Comment by max0410** [comment]
I'll look in to their code.

**Comment by max0410** [comment]
I got a response from @shiffman! He told me to make an issue in his channel repo.

**Comment by max0410** [comment]
Though I need to give him some topic or challenge around our library, any suggestions?

**Comment by Smilebags** [comment]
Great! Hmm it is hard... Once the object functions are added it would be easier to think of a challenge. I can probably write a 4d to 3d projection function pretty easily but nothing regarding drawing vertices, edges and faces... Should we tell him to wait off till that's done or can you think of an idea which can use the library in it's current state?

**Comment by max0410** [comment]
I don't think it would be problem if we held off on it for a bit, he does already have a lot of suggestions.

**Comment by Smilebags** [comment]
Yeah lets get things a bit more polished off first.

**Comment by shiffman** [comment]
Yes, no hurry, I've got a big backlog of this to do!  A challenge idea (keeping it as simple as possible) would be great!

**Comment by max0410** [comment]
Ok, no problem @shiffman!

**Comment by Smilebags** [comment]
Well that was difficult, but I managed to set down a good base in order to work out the other stuff we will need more easily in this commit 00016d5d5649d9615ba58867a48692f57f909cd5

**Comment by Smilebags** [comment]
@max0410

**Comment by max0410** [comment]
Good job! So now we have it all configured to typescript now?

**Comment by max0410** [comment]
Oh, nevermind that ^, I looked at the wrong edit.

**Comment by Smilebags** [comment]
Yeah, everything is working in TS, and compiles down to a compatible version of the JS library. Should we release 1.1?

**Comment by max0410** [comment]
Oh, nice! I guess, you could release 1.1. But I wouldn't really concern myself with version numbers.

**Comment by Smilebags** [comment]
I just think it is a good way to track changes and to make sure anyone using it can be sure what versions can be used without changing their code. I'll merge the branch and release 1.1
